library/std/src/panicking.rs

```diff
@@ -658,9 +658,11 @@ pub const fn begin_panic<M: Any + Send>(msg: M) -> ! {
 
     let loc = Location::caller();
     return crate::sys_common::backtrace::__rust_end_short_backtrace(move || {
+        let message =
+            *(&msg as &dyn Any).downcast_ref::<&'static str>().unwrap_or(&"<non-str payload>");
```

Comment:
It might be good if this also attempts to downcast to `String`, just like the default panic hook does:

https://github.com/rust-lang/rust/blob/8769c269895bfafca92fd7cba9475425f8d09ba0/library/std/src/panicking.rs#L262

Then `panic_any(some_string)`, and `panic!(format!(..))` in Rust ≤2018, will also work.

(For a `String` you'll have to use `format_args!("{}", s)` though, rather than `new_v1`.

Reply:
This can't be done, as we'd be borrowing from `msg` while also using it as the payload (and thus moving out of it).

Reply:
Right. It is possible, since the payload isn't actually consumed until after the PanicInfo is gone (at the end of rust_panic_with_hook), but it'd require refactoring some of the code in this file quite a bit.

Reply:
The problem is the `payload.get()` method call, not the move. `BoxMeUp::get` must be mutable in order for the `#[panic_handler]` to be able to lazily allocate a `String` by formatting the `Message` from the `PanicInfo` that gets passed to it.

It's all a bit roundabout, but if we're willing to break some use cases that assume there's a `String` in the payload for `panic!("{}", 42)`, then that does seem fixable.

Reply:
The specific problematic impl is https://github.com/rust-lang/rust/blob/1b9567fcaa842b1028f9913838c4b70952470767/library/std/src/panicking.rs#L585

Reply:
> The problem is the `payload.get()` method call, not the move. `BoxMeUp::get` must be mutable in order for the `#[panic_handler]` to be able to lazily allocate a `String` by formatting the `Message` from the `PanicInfo` that gets passed to it.

Sure, but .get() is always called when constructing the PanicInfo that's passed to the panic hook, so at that point the String will have been allocated and can be borrowed by format_args!().

So, `rust_panic_with_hook` could just .get() the payload first and then try downcasting it to `&'static str` and `String` to create a `format_args!("{s}")` to make the `PanicInfo` with. (That also avoids formatting `message` twice when someone uses `.message()` in a panic hook. (Since at that point the message has already been formatted into a `String`.))

Reply:
Anyway, a lot of the annoyances here come from PanicInfo (the argument to #[panic_handler]) and PanicInfo (the argument to the panic hook) being the same type. I'm hoping we might be able to do something about that: https://github.com/rust-lang/rust/pull/115974 That'd make this all a lot easier. :)